File: simplified-books-core-ng/src/main/java/org/nypl/simplified/books/book_database/BookDatabases.java
package org.nypl.simplified.books.book_database;

import org.nypl.simplified.books.accounts.AccountID;
import org.nypl.simplified.opds.core.OPDSJSONParser;
import org.nypl.simplified.opds.core.OPDSJSONParserType;
import org.nypl.simplified.opds.core.OPDSJSONSerializer;
import org.nypl.simplified.opds.core.OPDSJSONSerializerType;

import java.io.File;

public final class BookDatabases implements BookDatabaseFactoryType {

  private static final BookDatabases INSTANCE = new BookDatabases();

  public static BookDatabases get() {
    return INSTANCE;
  }

  private BookDatabases() {

  }

  @Override
  public BookDatabaseType openDatabase(
      final OPDSJSONParserType parser,
      final OPDSJSONSerializerType serializer,
      final AccountID owner,
      final File directory) throws BookDatabaseException {

    return BookDatabase.open(parser, serializer, owner, directory);
  }

  @Override
  public BookDatabaseType openDatabase(
      final AccountID owner,
      final File directory) throws BookDatabaseException {

    return BookDatabase.open(
        OPDSJSONParser.newParser(), OPDSJSONSerializer.newSerializer(), owner, directory);
  }
}

File: simplified-books-database/src/main/java/org/nypl/simplified/books/book_database/BookDatabases.kt
package org.nypl.simplified.books.book_database

import android.content.Context
import org.nypl.simplified.accounts.api.AccountID
import org.nypl.simplified.books.book_database.api.BookDatabaseException
import org.nypl.simplified.books.book_database.api.BookDatabaseFactoryType
import org.nypl.simplified.books.book_database.api.BookDatabaseType
import org.nypl.simplified.books.formats.api.BookFormatSupportType
import org.nypl.simplified.opds.core.OPDSJSONParser
import org.nypl.simplified.opds.core.OPDSJSONParserType
import org.nypl.simplified.opds.core.OPDSJSONSerializer
import org.nypl.simplified.opds.core.OPDSJSONSerializerType
import java.io.File

object BookDatabases : BookDatabaseFactoryType {

  @Throws(BookDatabaseException::class)
  override fun openDatabase(
    context: Context,
    parser: OPDSJSONParserType,
    serializer: OPDSJSONSerializerType,
    formats: BookFormatSupportType,
    owner: AccountID,
    directory: File
  ): BookDatabaseType {
    return BookDatabase.open(
      context = context,
      parser = parser,
      serializer = serializer,
      formats = formats,
      owner = owner,
      directory = directory
    )
  }

  @Throws(BookDatabaseException::class)
  override fun openDatabase(
    context: Context,
    formats: BookFormatSupportType,
    owner: AccountID,
    directory: File
  ): BookDatabaseType {
    return BookDatabase.open(
      context = context,
      parser = OPDSJSONParser.newParser(),
      serializer = OPDSJSONSerializer.newSerializer(),
      formats = formats,
      owner = owner,
      directory = directory
    )
  }
}
